### PR TITLE
KSBD: new source with flavour text, some code cleanup

### LINKED ISSUE #3505: [Feature Request] KSBD: add flavor texts
follow up to #1588

### PR DESCRIPTION
**Pull request by AriaMoradi**
closes #3505

### CODE REVIEW

**Comment by AriaMoradi** [comment]
![Screenshot_20200612-180520_Tachiyomi](https://user-images.githubusercontent.com/8719233/84508308-560f2280-acd7-11ea-9ad6-6b1fed3df766.jpg
Sample image

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt [hunk 1]
```diff
@@ -19,9 +19,9 @@ import rx.Observable
  *  @author Aria Moradi <aria.moradi007@gmail.com>
  */
 
-class KillSixBillionDemons : HttpSource() {
+open class KillSixBillionDemons : HttpSource() {
 
-    override val name = "KillSixBillionDemons"
+    override val name = "Kill Six Billion Demons"
```
Comment (CarlosEsco) [inline src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt:24]: Can't change the name now.  It fucks with the generatedID

    override val id by lazy {
        val key = "${name.toLowerCase()}/$lang/$versionId"
        val bytes = MessageDigest.getInstance("MD5").digest(key.toByteArray())
        (0..7).map { bytes[it].toLong() and 0xff shl 8 * (7 - it) }.reduce(Long::or) and Long.MAX_VALUE
    }

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import eu.kanade.tachiyomi.source.SourceFactory
+
+/**
+ *  @author Aria Moradi <aria.moradi007@gmail.com>
+ */
+
+class KillSixBillionDemonsFactory : SourceFactory {
+    override fun createSources(): List<Source> = listOf(
+        KillSixBillionDemons(),
+        KillSixBillionDemonsWithFlavourText()
```
Comment (CarlosEsco) [inline src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: Why is it two sources now?

**Comment by AriaMoradi** [comment]
There is a bug, I can read all of a chapter just fine, but trying to download a chapter fails.
It downloads all pages fine, but errors out before finishing
I checked the files, page 1 and 3 are downloaded and that's despite progress bar filling up to 100%

Screenshot_20200612-180230_Tachiyomi

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt (hunk 1 above)
Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt:24]: is it fine if I copy and paste the old id?
the old name was kinda wrong

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt (hunk 2 above)
Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: checkout #1588

**Comment by AriaMoradi** [comment]
@CarlosEsco it's not WIP though

**Comment by CarlosEsco** [comment]
You just posted there is a bug with download aka its WIP.

**Comment by AriaMoradi** [comment]
the bug mentioned is most likely with the app itself.
not using OKHTTP I presume.

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt (hunk 1 above)
Comment (CarlosEsco) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt:24]: yes

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt (hunk 2 above)
Comment (CarlosEsco) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: yes but having it a separate source means.  You have to add the entry twice into your library.  Is it not possible to have image 1 be from the url and image 2 be the "flavour text"

Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: 5520021760150818261

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt (hunk 1 above)
Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemons.kt:24]: original id = 5520021760150818261

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt (hunk 2 above)
Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: thats exatly how the new source is
some may prefer to read without flavor texts

**Comment by CarlosEsco** [comment]
I assume if you have issues downloading its something you are doing in your new class that either needs fixed, or a main app change.  So until the reason the download hangs is found out this will remain WIP

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt (hunk 2 above)
Comment (CarlosEsco) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: if the text is part of the story why would anyone want the non text version?  This just adds an extra extension for no real reason.

Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: still disagree
we need more opinions on this

Comment (CarlosEsco) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: So if you were making the xkcd extension from scratch you think having a version with and a version without the extra text is a good idea?

Comment (arkon) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: I don't really even agree with having sources that rely on text in a manga app.

Comment (CarlosEsco) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: agree with that too, i originally thought it was a few lines or paragraph.  This pages of text.

**Comment by AriaMoradi** [comment]
here's what happens when the app tries to download a chapter from this
```
2020-06-12 20:30:26.446 27837-28274/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/
2020-06-12 20:31:36.387 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/chapter/5-wielder-of-names/?order=ASC
2020-06-12 20:31:37.371 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/chapter/5-wielder-of-names/page/2/?order=ASC
2020-06-12 20:31:37.943 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/chapter/5-wielder-of-names/page/3/?order=ASC
2020-06-12 20:31:38.282 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: making pages
2020-06-12 20:31:38.296 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: returning pages
2020-06-12 20:31:38.353 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: returning GET for https://killsixbilliondemons.com/wp-content/uploads/2016/05/WON81.jpg)
2020-06-12 20:31:38.359 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/wp-content/uploads/2016/05/WON81.jpg
2020-06-12 20:31:40.654 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: returning GET for https://killsixbilliondemons.com/wp-content/uploads/2016/06/WON91.jpg)
2020-06-12 20:31:40.657 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/wp-content/uploads/2016/06/WON91.jpg
2020-06-12 20:31:42.071 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: returning GET for https://killsixbilliondemons.com/wp-content/uploads/2016/07/WON101.jpg)
2020-06-12 20:31:42.074 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/wp-content/uploads/2016/07/WON101.jpg
2020-06-12 20:31:43.574 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-81/
2020-06-12 20:31:43.924 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-81/
2020-06-12 20:31:44.241 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-82/
2020-06-12 20:31:44.647 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-82/
2020-06-12 20:31:44.958 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-83/
2020-06-12 20:31:45.378 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-83/
2020-06-12 20:31:45.669 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-84/
2020-06-12 20:31:45.933 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-84/
2020-06-12 20:31:46.622 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-85/
2020-06-12 20:31:46.884 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-85/
2020-06-12 20:31:47.399 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-86-ball/
2020-06-12 20:31:47.694 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-86-ball/
2020-06-12 20:31:48.017 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-87/
2020-06-12 20:31:48.303 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-87/
2020-06-12 20:31:48.570 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-88/
2020-06-12 20:31:49.145 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-88/
2020-06-12 20:31:49.439 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-89/
2020-06-12 20:31:49.741 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-89/
2020-06-12 20:31:51.722 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-90/
2020-06-12 20:31:51.984 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-90/
2020-06-12 20:31:54.285 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-91/
2020-06-12 20:31:54.652 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-91/
2020-06-12 20:31:56.742 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-92-tree-of-woe/
2020-06-12 20:31:57.019 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-92-tree-of-woe/
2020-06-12 20:31:57.380 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-93/
2020-06-12 20:31:57.657 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-93/
2020-06-12 20:32:00.748 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-94/
2020-06-12 20:32:01.846 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-94/
2020-06-12 20:32:08.235 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-95/
2020-06-12 20:32:09.114 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-95/
2020-06-12 20:32:09.646 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-96/
2020-06-12 20:32:10.012 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-96/
2020-06-12 20:32:10.534 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-97/
2020-06-12 20:32:10.792 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-97/
2020-06-12 20:32:11.080 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-98/
2020-06-12 20:32:11.345 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-98/
2020-06-12 20:32:11.623 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-99/
2020-06-12 20:32:12.040 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-99/
2020-06-12 20:32:12.461 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-100/
2020-06-12 20:32:12.887 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-100/
2020-06-12 20:32:13.472 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-101/
2020-06-12 20:32:13.871 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-101/
2020-06-12 20:32:14.214 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: going to try intercepting https://killsixbilliondemons.com/comic/wielder-of-names-5-102/
2020-06-12 20:32:14.621 27837-28284/eu.kanade.tachiyomi.debug D/ksbd: flavour page image url is going to be made from: https://killsixbilliondemons.com/comic/wielder-of-names-5-102/
```
the app calls `imageRequest()` a few times(2 or 3) for the first pages, and GET is done and files are saved, then just gets all text image urls and then fails.
intercepting messeges are from DataImageInterceptor which essentially documents every http request the app sends.
it's strange for me that the app won't even try to call `imageRequest()` for text image urls.

File: src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt (hunk 2 above)
Comment (AriaMoradi) [reply src/en/killsixbilliondemons/src/eu/kanade/tachiyomi/extension/en/killsixbilliondemons/KillSixBillionDemonsFactory.kt:13]: it doesn't rely on text, the printed version doesn't have them either, it's like a regular comic.
the text is like some extra lore from the universe, pieces of a fictional bible.

**Comment by AriaMoradi** [comment]
the same similar problem is preset with `Manga Ship` which uses `DataImageInterceptor` for all it's pages

**Comment by SnakeDoc83** [comment]
Can't speak to why your implementation doesn't download, but I got it to download for Manga Ship.

**Comment by AriaMoradi** [comment]
after reading the code for the fifth time, found the but, it was propagated from the original code
anyways, it's fixed now.
ganna remove the WIP tag
and lets get back to taking about whether this code is worth merging or not...

**Comment by arkon** [comment]
Anything that's reliant on text is a no from me. Xkcd is sort of pushing it, but it's alt text that isn't actually _required_. This doesn't really seem like that, although I don't read it.

**Comment by AriaMoradi** [comment]
> It adds lore to the story, they are like pieces of a seudo-bible of that universe. It's not fundamental to understand the story, but it certainly adds to the experience.

_Originally posted by @emidelgado01 in https://github.com/inorichi/tachiyomi-extensions/issues/1588#issuecomment-640916540_

**Comment by arkon** [comment]
Closing this since it falls out of Tachiyomi's intended usage.

**Pull request closed without merging**